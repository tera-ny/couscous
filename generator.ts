import { FunctionDeclaration } from "ts_morph";
import type { Route } from "./type.ts";

export const addEntry = (route: Route, constructor: FunctionDeclaration) => {
  constructor.addOverload({
    parameters: [{ name: "identity", type: `"${route.identity}"` }],
    returnType: "void",
  });
};

const test = {
  "/users/[id]/": () => {},
};
test["/users/[id]/"]();
