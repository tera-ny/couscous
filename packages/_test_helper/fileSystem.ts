import {
  Project,
  SourceFile,
  ManipulationSettings,
  IndentationText,
} from "https://deno.land/x/ts_morph@14.0.0/mod.ts";

export const withDir = async (
  cb: (tempDir: string) => void | Promise<void>
) => {
  const tempDir = await Deno.makeTempDir({
    prefix: "deno_couscous_test_",
  });
  await cb(tempDir);
  await Deno.remove(tempDir, { recursive: true });
};

export const withSource = (
  cb: (tempSrc: SourceFile) => void | Promise<void>,
  manipulationSettings: Partial<ManipulationSettings> = {
    indentationText: IndentationText.TwoSpaces,
  }
) =>
  withDir(async (tempDir) => {
    const tempFile = await Deno.makeTempFile({ dir: tempDir });
    const project = new Project({
      libFolderPath: tempDir,
      manipulationSettings,
    });
    const source = project.addSourceFileAtPath(tempFile);
    await cb(source);
  });

type TestParameter = Deno.TestContext | TestCallback;
type TestCallback = (
  tempSrc: SourceFile,
  context: Deno.TestContext
) => void | Promise<void>;

export function test(name: string, cb: TestCallback): void;
export function test(
  name: string,
  context: Deno.TestContext,
  cb: TestCallback
): Promise<boolean>;
export function test(name: string, ...args: TestParameter[]) {
  if (typeof args[0] === "function") {
    const cb = args[0] as TestCallback;
    Deno.test(name, { permissions: { read: true, write: true } }, (t) =>
      withSource((source) => cb(source, t))
    );
  } else {
    const context = args[0] as Deno.TestContext;
    const cb = args[1] as TestCallback;
    return context.step(name, (t) => withSource((source) => cb(source, t)));
  }
}
