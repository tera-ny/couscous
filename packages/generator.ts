import { FunctionDeclaration } from "https://deno.land/x/ts_morph@14.0.0/mod.ts";
import type { Route } from "./type.ts";

export const addEntryOverloads = (
  routes: Route[],
  constructor: FunctionDeclaration
) => {
  constructor.addOverloads(
    routes.map((route) => ({
      parameters: [
        { name: "identity", type: `"${route.identity}"` },
        ...route.query.map((key) => ({ name: key, type: "string" })),
        {
          name: "option",
          hasQuestionToken: true,
          type: "RouteOption",
        },
      ],
      returnType: "string",
    }))
  );
};

export const addRoutesHandler = (
  routes: Route[],
  routesHandler: FunctionDeclaration
) => {
  routesHandler.addStatements((writer) => {
    writer.writeLine("let path: string;");
    writer.writeLine("let index: number;");
    writer.write("switch (identity)").block(() => {
      routes.forEach((route) => {
        writer.writeLine(`case "${route.identity}":`);
        writer.block(() => {
          route.query.forEach((key, index) => {
            writer.writeLine(`const ${key} = args[${index}];`);
          });
          writer.writeLine("path = `" + route.template + "`;");
          const optionIndex = route.query.length;
          writer.writeLine(`index = ${optionIndex};`);
          writer.write("break;");
        });
      });
    });
    writer.writeLine(
      "const option = args[index] as (RouteOption | undefined);"
    );
    writer.writeLine(
      "const searchParams = new URLSearchParams(option?.query ?? {});"
    );
    writer.writeLine(
      "return `${path}${toSearch(searchParams)}${option?.hash}`"
    );
  });
};
