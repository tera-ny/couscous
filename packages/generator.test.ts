import { StructureKind } from "https://deno.land/x/ts_morph@14.0.0/mod.ts";
import { test } from "./_test_helper/fileSystem.ts";
import { assertSnapshot } from "https://deno.land/std@0.136.0/testing/snapshot.ts";
import { addEntryOverloads } from "./generator.ts";

test("addEntryOverloads with snapshot", async (source, ctx) => {
  const constructor = source.addFunction({
    name: "route",
    kind: StructureKind.Function,
    statements: 'return ""',
  });
  addEntryOverloads(
    [
      { identity: "/fuga", template: "/fuga/", params: [] },
      // single parameter
      {
        identity: "/hoge/[hoge]",
        template: "/hoge/${hoge}",
        params: [{ name: "hoge", type: "single" }],
      },
      // rest parameter
      {
        identity: "/piyo/[...slug]",
        template: '/piyo/${slug.join("/")}',
        params: [{ name: "slug", type: "rest", isOptional: false }],
      },
      {
        identity: "/foo/[foo]/[...bar]",
        template: '/foo/${foo}/${bar.join("/")}',
        params: [
          { name: "foo", type: "single" },
          { name: "bar", type: "rest", isOptional: false },
        ],
      },
      // optional rest parameter
      {
        identity: "/foo/[foo]/[[...piyo]]",
        template: '/foo/${foo}/${piyo.join("/")}',
        params: [
          { name: "foo", type: "single" },
          { name: "bar", type: "rest", isOptional: true },
        ],
      },
    ],
    constructor
  );
  await source.save();
  const target = await Deno.readTextFile(source.getFilePath());
  await assertSnapshot(ctx, target);
});
