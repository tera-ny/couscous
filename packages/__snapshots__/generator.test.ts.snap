export const snapshot = {};

snapshot[`addEntryOverloads with snapshot 1`] = `'function route(identity: "/fuga", option?: RouteOption): string;\\nfunction route(identity: "/hoge/[hoge]", hoge: string, option?: RouteOption): string;\\nfunction route(identity: "/piyo/[...slug]", slug: [string, ...string[]], option?: RouteOption): string;\\nfunction route(identity: "/foo/[foo]/[...bar]", foo: string, bar: [string, ...string[]], option?: RouteOption): string;\\nfunction route(identity: "/foo/[foo]/[[...piyo]]", foo: string, bar: string[], option?: RouteOption): string;\\nfunction route() {\\n  return ""\\n}\\n'`;
